qutip/krylovsolve.py

```diff
@@ -0,0 +1,690 @@
+            options,
+            res,
+            psi0,
+            expt_callback,
+            e_0,
+            n_expt_op,
+            e_ops,
+            e_m_0,
+        )
+
+    if (options.store_final_state) and (not options.store_states):
+        res.states = [psi0]
+
+    progress_bar.update(1)
+    progress_bar.finished()
+    return res
+
+
+def _happy_breakdown(
+    tlist, options, res, psi0, expt_callback, e_0, n_expt_op, e_ops, e_m_0
+):
+    """
+    Dummy evolves the system if a happy breakdown of an eigenstate occurs.
+    """
+    for i in range(1, len(tlist)):
+        if options.store_states:
+            res.states.append(psi0)
+        if expt_callback:
+            res.expect.append(e_0)
+
+        for m in range(n_expt_op):
+            op = e_ops[m]
+            res.expect[m][i] = e_m_0[m]
+    return res
+
+
+def _optimize_lanczos_timestep_size(T, krylov_basis, tlist, tol):
+    """
+    Solves the equation defined to optimize the number of Lanczos
+    iterations to be performed inside Krylov's algorithm.
+    """
+
+    f = _lanczos_error_equation_to_optimize_delta_t(
+        T,
+        krylov_basis=krylov_basis,
+        t0=tlist[0],
+        tf=tlist[-1],
+        target_tolerance=tol,
+    )
+    bracket = [tlist[0], tlist[-1]]
+
+    if (np.sign(f(tlist[0])) == -1) and (np.sign(f(tlist[-1])) == -1):
+        delta_t = tlist[-1] - tlist[0]
+        return delta_t
+    else:
+        sol = root_scalar(f=f, bracket=bracket, method="brentq", xtol=tol)
+        if sol.converged:
+            delta_t = sol.root
+            return delta_t
+        else:
+            raise Exception(
+                "Method did not converge, try increasing 'krylov_dim', "
+                "taking a lesser final time 'tlist[-1]' or decreasing the "
+                "tolerance via Options().atol. "
+                "If nothing works, this problem might not be suitable for "
+                "Krylov or a deeper analysis might be required."
+            )
+
+
+def _lanczos_error_equation_to_optimize_delta_t(
+    T, krylov_basis, t0, tf, target_tolerance
+):
+    """
+    Function to optimize in order to obtain the optimal number of
+    Lanczos algorithm iterations, governed by the optimal timestep size between
+    Lanczos iteractions.
+    """
+    eigenvalues1, eigenvectors1 = eigh(T[0:, 0:])
+    U1 = np.matmul(krylov_basis[0:, 0:].T, eigenvectors1)
+    e01 = eigenvectors1.conj().T[:, 0]
+
+    eigenvalues2, eigenvectors2 = eigh(T[0:-1, 0: T.shape[1] - 1])
+    U2 = np.matmul(krylov_basis[0:-1, :].T, eigenvectors2)
+    e02 = eigenvectors2.conj().T[:, 0]
+
+    def f(t):
+        delta_t = -1j * (t - t0)
+
+        aux1 = np.multiply(np.exp(delta_t * eigenvalues1), e01)
+        psi1 = np.matmul(U1, aux1)
+
+        aux2 = np.multiply(np.exp(delta_t * eigenvalues2), e02)
+        psi2 = np.matmul(U2, aux2)
+
+        error = np.linalg.norm(psi1 - psi2)
+
+        steps = 1 if t == t0 else max(1, (tf - t0) // (t - t0))
```

Comment:
Hello Eric, most of what's been said here I've taken into consideration. About the <code>dt<<1</code> situation, one possibility could be to have a user parameter from the options class. Something like <code>Options.nsteps</code> could stand for the maximum allowed of lanczos iterations, so it can be controlled by the user. Let me know what you think.

Regarding the small subsystems yes, krylov is not designed for those. I could return sesolve, but I think it might be conflictive if someone is doing krylov specific research. Maybe in the future it can have an option to either rely on sesolve if krylov failse, if not, then simply raise an exception.

Reply:
Yes using `Options.nsteps` would be a good options to limit very small `dt`.

For small systems, it could fail cleanly or use sesolve, but it should not fill then ram and hang.

Reply:
For the time being I think it's preferrable to raise an Exception, it can be confusing if it changes the algorithm without proper documentation. Also, behaviour between krylov and sesolve is different for an empty tlist, so it could be a bit annoying.